cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricSampleAggregator.java

```diff
@@ -0,0 +1,405 @@
+
+  private final ConcurrentMap<E, RawMetricValues> _rawMetrics;
+  private final MetricSampleAggregatorState<G, E> _aggregatorState;
+  private final ReentrantLock _windowRollingLock;
+
+  protected final ConcurrentMap<E, E> _identityEntityMap;
+  protected final int _numWindows;
+  protected final int _minSamplesPerWindow;
+  protected final int _numWindowsToKeep;
+  protected final long _windowMs;
+  protected final int _maxAllowedImputations;
+  protected final MetricDef _metricDef;
+  
+  private volatile long _currentWindowIndex;
+  private volatile long _oldestWindowIndex;
+
+  /**
+   * Construct the abstract metric sample aggregator.
+   *
+   * @param numWindows the number of snapshot windows needed.
+   * @param windowMs the size of each snapshot window in milliseconds
+   * @param minSamplesPerWindow minimum samples per snapshot window.
+   * @param maxAllowedImputations the maximum allowed number of imputations for an entity if some windows miss data.
+   * @param metricDef metric definitions.
+   */
+  public MetricSampleAggregator(int numWindows,
+                                long windowMs,
+                                int minSamplesPerWindow,
+                                int maxAllowedImputations,
+                                MetricDef metricDef) {
+    super(0);
+    _identityEntityMap = new ConcurrentHashMap<>();
+    _rawMetrics = new ConcurrentHashMap<>();
+    _numWindows = numWindows;
+    _windowMs = windowMs;
+    // We keep one more window for the active window.
+    _numWindowsToKeep = _numWindows + 1;
+    _minSamplesPerWindow = minSamplesPerWindow;
+    _windowRollingLock = new ReentrantLock();
+    _maxAllowedImputations = maxAllowedImputations;
+    _metricDef = metricDef;
+    _aggregatorState = new MetricSampleAggregatorState<>(_generation.get(), _windowMs);
+    _oldestWindowIndex = 0L;
+    _currentWindowIndex = 0L;
+  }
+
+  /**
+   * Add a sample to the metric aggregator.
+   *
+   * @param sample The metric sample to add.
+   *
+   * @return true if the sample is accepted, false if the sample is ignored.
+   */
+  public boolean addSample(MetricSample<G, E> sample) {
+    if (!sample.isValid(_metricDef)) {
+      LOG.warn("The metric sample is discarded due to missing metrics. Sample: {}", sample);
+      return false;
+    }
+    long windowIndex = windowIndex(sample.sampleTime());
+    // Skip the samples that are too old.
+    if (windowIndex < _oldestWindowIndex) {
+      return false;
+    }
+    boolean newWindowRolledOut = rollOutNewWindowIfNeeded(windowIndex);
+    RawMetricValues rawMetricValues =
+        _rawMetrics.computeIfAbsent(identity(sample.entity()), k -> {
+          RawMetricValues rawValues = new RawMetricValues(_numWindowsToKeep, _minSamplesPerWindow);
+          rawValues.updateOldestWindowIndex(_oldestWindowIndex);
+          return rawValues;
+        });
+    LOG.trace("Adding sample {} to window index {}", sample, windowIndex);
```

Comment:
Even at trace level this seems like it might be called lots of times.  If this log statement needs to be kept then perhaps it can be in a if (LOG.isTraceEnabled()) { block }

Reply:
This is what slf4j does for you.